Added generic Python asyncio generator script.

olitheolix opened the pull request:
This is similar to the existing `python-asyncio.sh` script but does not hard code the name of the client library (`kubernetes_asyncio`). Instead, it uses the `PACKAGE_NAME` variable from the settings file as explained in the Readme to this repository.

This will allow others to create their own client libraries more easily.

olitheolix commented:
@k8s-ci-robot

tomplus commented:
I'm not sure if it's place for "generic" clients. This special repo `gen` is dedicated to client libraries which are part of https://github.com/kubernetes-client/. I hope `kubernetes_asyncio` will be here soon.

olitheolix commented:
> I hope kubernetes_asyncio will be here soon.
And I hope `aiokubernetes` will be too. Shall I rename the file to `python-aiokubernetes.sh`?

brendandburns commented:
Can we just modify the existing script rather than introduce a new one?

I don't mind having scripts that are used for non `kubernetes-client` libraries as long as they don't make this repo harder to maintain, and in this case, making this script a little more generic (with appropriate defaults) seems fine with me.

olitheolix commented:
I did not modify `python-asyncio.sh` because it is customised for `kubernetes_asyncio` and might break it.

Instead, I copied it and:
* changed the `SWAGGER_CODEGEN_COMMIT` to an official version tag,
* replaced the hard coded `kubernetes_asyncio` with the shell variable from the `settings` scripts,
* removed the (also custom made for kubernetes_asyncio) patches.

This is as generic as I can think of making it right now. It also produces a working client.

All my `aiokubernetes` specific patches and modifications happen downstream where I can maintain them myself without having to bother anyone in this `gen` repository.

I can modify the PR to change `python-asyncio.sh` and delete `python-asyncio-rest.py.patch` but that may cause problems for `kubernetes_asyncio`. Can we keep both versions, at least for now, to reduce the risk of breaking anything for anyone?

brendandburns commented:
I don't want lots of slightly different scripts. I think you can add conditionalization to make this work correctly with a single `python-asyncio.sh` script.

brendandburns commented:
Sorry, this isn't quite what I hoped, the if/else should minimize the differences between the files, rather than just have a giant if/else across the two blocks.

There are many environment variables (e.g. CLEANUP_DIRS) which are shared in the two blocks and shouldn't be duplicated.

Please refactor to minimize differences.

Thanks (and apologies for not being clear before)

olitheolix commented:
Is this what you had in mind?

brendandburns commented:
Yep, that looks much better, thanks!

@tomplus can you validate that `python_asyncio` generates ok with this PR?

Thanks!

olitheolix commented on openapi/python-asyncio.sh:
```diff
@@ -44,34 +44,49 @@ popd > /dev/null
-find "${OUTPUT_DIR}/client/" -type f -name \*.py -exec sed -i 's/import client\./import kubernetes_asyncio.client./g' {} +
-find "${OUTPUT_DIR}/client/" -type f -name \*.py -exec sed -i 's/from client/from kubernetes_asyncio.client/g' {} +
-find "${OUTPUT_DIR}/client/" -type f -name \*.py -exec sed -i 's/getattr(client\.models/getattr(kubernetes_asyncio.client.models/g' {} +
+find "${OUTPUT_DIR}/test" -type f -name \*.py -exec sed -i "s/\\bclient/${PACKAGE_NAME}.client/g" {} +
```
Not sure what you want me to change here because this line is a verbatim copy of the one in master.

olitheolix commented on openapi/python-asyncio.sh:
```diff
@@ -44,34 +44,49 @@ popd > /dev/null
-CLIENT_LANGUAGE=python-asyncio; \
-CLEANUP_DIRS=(client/apis client/models docs test); \
+# Client specific Swagger branch to use.
+if [ ${PACKAGE_NAME} == "kubernetes_asyncio" ]; then
```
Fixed.

brendandburns commented:
@tomplus can we get a glance here to see if this does the right things?

Thanks!

tomplus commented:
@brendandburns Thanks for asking. I tested it and it works for me.

/lgtm

brendandburns commented:
/approve

brendandburns commented:
@olitheolix this is finally merging :) thanks for your patience.
